Move power subsystem to new power states

ceolin opened the pull request:
- Move power states and residency info to DTS
- Remove old power states

Still missing:

- some platform information in DTS
- Get rid of weak functions and move to new API proposed in another PR
- Use new constraint API
- Constraints don't apply to substate.

Commit:
power: Move pm subsystem to new power states

Migrate the whole pm subsystem to use new power states information
from power_state.h and get states and residency properties from
device tree.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
boards: cc1352r1_launchxl: Add idle states in dts

Add information about this board idle states in its dts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
boards: mec1501modular_assy6885: Add idle states in dts

Add idle states info in this board dts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
dts: power: Change pm state properties

Making it looks more Linux alike

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
boards: cc1352r_sensortag: Add idle states in dts

Add information about this board idle states in its dts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
boards: cc26x2r1_launchxl: Add idle states in dts

Add information about this board idle states in its dts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
boards: mec15xxevb_assy6853: Add idle states in dts

Add information about this board idle states in its dts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
dts: mec1501hsz: Add cpu label

Add label to facilitate add cpu-idle-states later.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: dts: Return empty list when the are no power states

Return an empty list when information about power states is not
defined in the device tree of an specific target.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Remove power management conditionals from code

Remove conditionals (PM_DEEP_SLEEP_STATES and PM_SLEEP_STATES) from
power management code. Now these features are always available when
power management is enabled.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Create necessary function stubs

Power management require stubs for pm_power_state_set() and
pm_power_state_exit_post_ops().

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Remove residency and states from Kconfig

Residency time and power states are defined using device tree now.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Add substate-id property for a power state

There are platforms that have multiple states that maps to a specific
Zephyr power state. To accommodate this sort of situation this commit
adds an additional property to a power state that can be used by the
platform.

The power state now consists of two properties, a category and a
substate-id. The former property is the current power state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Use substate-id in pm subsystem

Change subsystem to use struct pm_state with substate-id instead of
using only the power state category.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on samples/boards/mec15xxevb_assy6853/power_management/src/power_mgmt.c:
```diff
@@ -18,7 +18,7 @@ LOG_MODULE_REGISTER(pwrmgmt_test);
 #define TASK_STACK_SIZE           1024ul
 #define PRIORITY                  K_PRIO_COOP(5)
-/* Sleep time should be lower than CONFIG_PM_MIN_RESIDENCY_SLEEP_1 */
+/* Sleep time should be lower than SUSPEND_TO_IDLE state */
```
need to reword this, previous it was talking about residency, now it is talking about state.

mbolivar-nordic commented on include/power/power.h:
```diff
@@ -53,24 +53,24 @@ struct pm_notifier {
 	 * for power state entry.
 	 */
-	void (*state_entry)(enum pm_state state);
+	void (*state_entry)(struct pm_state state);
```
Various of these changes ought to be in the previous commit for bisectability, no?

mbolivar-nordic reviewed (commented):
No comments on the SoC specific parts, but I did notice a few things. Mostly related to docstrings and DT.

pabigot reviewed (commented):
Based on discussion in #31162 I think we need to revisit what a devicetree representation of a power state is, particularly to allow SoC-specific substates to be distinguishable, but also recognizing that `idle-state` may not be the right terminology.

ceolin commented on include/power/power_state.h:
```diff
@@ -104,11 +104,45 @@ enum pm_state {
+ * id allows a platform maps multiple specific power states to a Zephyr
+ * power management category.
+ */
+struct pm_state{
```
thanks, have fixed it.

pabigot commented:
> Based on discussion in #31162 I think we need to revisit what a devicetree representation of a power state is, particularly to allow SoC-specific substates to be distinguishable

I now see the substate properties are later in the PR, which is fine; please just move the commit that adds the property up near the other changes to the binding.

> but also recognizing that `idle-state` may not be the right terminology.

I believe we've agreed in the meeting that `power-state` is more accurate for the way Zephyr uses these.

ceolin commented:
> > Based on discussion in #31162 I think we need to revisit what a devicetree representation of a power state is, particularly to allow SoC-specific substates to be distinguishable
> 
> I now see the substate properties are later in the PR, which is fine; please just move the commit that adds the property up near the other changes to the binding.
> 
> > but also recognizing that `idle-state` may not be the right terminology.
> 
> I believe we've agreed in the meeting that `power-state` is more accurate for the way Zephyr uses these.

Indeed, but we agreed in wait others feedback. So, I'll address all other comments and wait a little bit more this one.

Commit:
soc: efm32wg: Select SOC_GECKO_CORE if pm is enabled

Power management requires core interrupt handling.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on boards/arm/cc1352r1_launchxl/cc1352r1_launchxl.dts:
```diff
@@ -59,10 +59,25 @@
 		};
 	};
+
+	idle-states {
```
So, in [Linux](https://www.kernel.org/doc/html/v5.0/admin-guide/pm/cpuidle.html)...

```
Modern processors are generally able to enter states in which the execution of a program is suspended and instructions belonging to it are not fetched from memory or executed. Those states are the idle states of the processor.

Since part of the processor hardware is not used in idle states, entering them generally allows power drawn by the processor to be reduced and, in consequence, it is an opportunity to save energy.

CPU idle time management is an energy-efficiency feature concerned about using the idle states of processors for this purpose.
```

pabigot replied:
Yes.  What does that have to do with our need to describe power states that aren't idle states?

nashif replied:
> Yes. What does that have to do with our need to describe power states that aren't idle states?

A lot. idle states here applies to all states that can be transitioned to when the cpu is idle, it is not referring to the idle power state.

pabigot replied:
OK.  So then we need a `power-states` that provides all the states, regardless of whether they can be used when transitioning to idle?  Because the power management infrastructure should allow us to move between them, not just pop into and out of a generic idle condition.

pabigot replied:
Another way: We know which of the power states are capable of being entered when the CPU is idle.  It's the ones that aren't `PM_STATE_ACTIVE` and `PM_STATE_RUNTIME_IDLE`.   That's indicated within each state by (currently) `idle-state-name`.

erwango commented on soc/arm/st_stm32/stm32l4/power.c:
```diff
@@ -17,12 +17,10 @@
-#ifdef CONFIG_PM_SLEEP_STATES
-#ifdef CONFIG_HAS_POWER_STATE_SLEEP_1
-	case POWER_STATE_SLEEP_1:
+	case PM_STATE_RUNTIME_IDLE:
```
PM_STATE_SUSPEND_TO_IDLE according to https://github.com/zephyrproject-rtos/zephyr/issues/31162#issuecomment-756818940
Not sure to see the interest of using a new nomenclature if this is to use misleading names

ceolin replied:
done

nashif commented:
> I believe we've agreed in the meeting that `power-state` is more accurate for the way Zephyr uses these.

we agreed to not change anything until we finalize the terminology, understand the background for idle-states naming in Linux and get more input from others. @galak, what is your input here.

galak commented:
> > 
> 
> we agreed to not change anything until we finalize the terminology, understand the background for idle-states naming in Linux and get more input from others. @galak, what is your input here.

I agree we should try and keep with the Linux binding names for now.

Commit:
soc: nordic: Disable PM for SOC_NRF5340_CPUNET_QKAA

This targets was previously enabling PM but the code was doing nothing
because the logic was under an unsatisfied ifdef condition.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

pabigot commented:
dev-review 2021-01-21: Agreed that it's OK to rename `idle-states` to `power-states` and recognize that we're using this to describe any system level power state, even if it isn't an idle state.

Also that this does *not* mean we're discarding using the Linux devicetree and API as a general architecture framework.  Only where names or behaviors in that are clearly not suitable to Zephyr's needs should we deviate from this.

The above statement applies generally to the design of the power management infrastructure, and should be incorporated into any documentation of that architecture that might exist.

@galak @carlescufi @nashif please comment if this does not reflect your understanding of the conclusion.

pabigot commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -12,15 +12,15 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
- * POWER_STATE_SLEEP_3: EM3 Stop
+ * PM_STATE_RUNTIME_IDLE: EM1 Sleep
+ * PM_STATE_SUSPEND_TO_IDLE: EM2 Deep Sleep
+ * PM_STATE_STANDBY: EM3 Stop
```
This doesn't seem to match https://github.com/zephyrproject-rtos/zephyr/issues/31162#issuecomment-757337571, but I guess that can be fixed later.

ceolin replied:
The comment was not super conclusive and was not clear about EM2 and EM3, also right now is not clear how to distinguish between them, so just keep this way for clarity sake.

pabigot reviewed (commented):
I don't think this is quite right yet but I also don't think we know enough to say what's right, so I have no objections or changes requested.  Will wait until CI passes for a final check before approving.

@erwango please revisit your nak.

pabigot reviewed (approved):
Remaining build failures in https://buildkite.com/zephyr/zephyr/builds/18862 don't seem related to anything I'm concerned with.

Later: They're  not: see #31530

erwango reviewed (approved):
Like it!
